### PR TITLE
[luci] Weight quantized-dequantized module about channel-wise quant uint8

### PR DESCRIPTION
**Pull request by meejeong**
Weight quantized-dequantized module about channel-wise quant uint8
- implement to distinguish which convolution operators (```get_channel_dim_index``` function)
- quantize & dequantize weights per channel
- datatype : uint8
- asymmetric quantization

ONE-DCO-1.0-Signed-off-by: mia park <mia.park@samsung.com>

Related to: #696, https://github.sec.samsung.net/STAR/nnapps/issues/379

### CODE REVIEW

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 1]
```diff
@@ -70,6 +70,183 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
 
+void get_channel_dim_index(CircleConst *node, int *len_per_dim, int &channel_dim_index)
+{
+  auto succs = loco::succs(node);
+  if (succs.size() != 1) // assume weights is used by only one node
+    return;
+
+  for (auto out : succs)
+  {
+
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:81]: Please remove extra line

**Comment by mhs4670go** [comment]
@nnfw-bot test nobody-checker

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 2]
```diff
@@ -70,6 +70,183 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+int cal_offset(int *len_per_dim, int *indices)
+{
+  return indices[0] * len_per_dim[1] * len_per_dim[2] * len_per_dim[3] +
+         indices[1] * len_per_dim[2] * len_per_dim[3] + indices[2] * len_per_dim[3] + indices[3];
+}
+
+void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max)
+{
+
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:123]: please remove this extra line

**Comment by mhs4670go** [comment]
@meejeong Could you change the commit message to concated(?) version?
e.g 
- a b c
- d e f
-> a b c and d e f

**Comment by meejeong** [comment]
> @meejeong Could you change the commit message to concated(?) version?
> e.g
> - a b c
> - d e f
> -> a b c and d e f

I applied it. thank you :-)

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 3]
```diff
@@ -70,10 +69,182 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+int cal_offset(int *len_per_dim, int *indices)
+{
+  return indices[0] * len_per_dim[1] * len_per_dim[2] * len_per_dim[3] +
+         indices[1] * len_per_dim[2] * len_per_dim[3] + indices[2] * len_per_dim[3] + indices[3];
+}
+
+void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max)
+{
+  int len_per_dim[4];
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:121]: there seems to be some condition about [4]. this may cause problems.
can you add at least some conditional check with assert()?
and if possible it would be better to check those condition for number 4.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 4]
```diff
@@ -70,10 +69,182 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+  int indices[4];
+  int channel_dim_index;
+  int size;
+
+  get_channel_dim_index(node, len_per_dim, channel_dim_index);
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:126]: there is no return/check of `get_channel_dim_index()` and this leads uninitalized `len_per_dim` that is used below.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 5]
```diff
@@ -40,7 +40,6 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
   const double qmax_double = kMaxScale;
   const double rmin = std::fmin(0, min);
   const double rmax = std::fmax(0, max);
-
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:43]: this change doesn't seem to be related with this PR

**Comment by mhs4670go** [comment]
:cry:
Maybe make it one line..? `implement to choose Convolution operators with quantize & dequantize weights per channel` or sth.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 6]
```diff
@@ -201,29 +373,58 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
...
-        float min = std::numeric_limits<float>::max();
-        float max = std::numeric_limits<float>::min();
-        for (uint32_t i = 0; i < circle_const->size<loco::DataType::FLOAT32>(); i++)
+        else
         {
-          auto data = circle_const->at<loco::DataType::FLOAT32>(i);
-          min = data < min ? data : min;
-          max = data > max ? data : max;
+          float min = std::numeric_limits<float>::max();
+          float max = std::numeric_limits<float>::min();
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:406]: Please use numeric_limits<float>::lowest() instead of numeric_limits<float>::min()

Comment (meejeong) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:406]: @jinevening , I wonder if there is a difference?

**Comment by mhs4670go** [comment]
[title]
`implement to choose Convolution operators with quantize & dequantize weights per channel`
title is necessary;ㅅ;

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 7]
```diff
@@ -113,7 +284,6 @@ void wquant_with_minmax_per_layer(CircleConst *node, float min, float max, float
 
 void wdequant_with_minmax_per_layer(CircleConst *node, float scaling_factor, float nudged_min)
 {
-
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:116]: I think this change was from previous commit with addition line, but added cause of my comment.
As this is not related with this PR, can you please separate this change to another PR ?

Comment (meejeong) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:116]: thank you for your advice. I will apply it.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp (hunk 6 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:406]: > @jinevening , I wonder if there is a difference?

`numeric_limits<float>::min()` is the smallest **positive** value. 

`numeric_limits<float>::lowest()` is the smallest **negative** value.

I guessed the intention of the code is to use the smallest negative value.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 8]
```diff
@@ -70,10 +69,182 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+      channel_dim_index = 2;
+    }
+    else if (fc != nullptr && fc->weights() == node)
+    {
+      len_per_dim[0] = 1;
+      len_per_dim[1] = 1;
+      len_per_dim[2] = node->dim(0).value();
+      len_per_dim[3] = node->dim(1).value();
+      channel_dim_index = 3;
+    }
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:109]: Please add an assertion to ensure that this function is called for the Ops supporting channel-wise quant.
E.g.,
```
else
{
  assert(false);
}
```

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 9]
```diff
@@ -201,29 +373,58 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
       {
         auto circle_const = loco::must_cast<luci::CircleConst *>(circle_node);
 
+        // Find min/max per channel-wise
+        if (granularity)
```
Comment (jinevening) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:377]: Can you clarify the condition? E.g., granularity == QuantizationGranularity::ChannelWise

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 10]
```diff
@@ -70,6 +70,193 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+
+void wdequant_per_channel(CircleConst *node, std::vector<float> &scaling_factor,
+                          std::vector<float> &nudged_min)
+{
+  uint32_t size = node->size<loco::DataType::U8>();
+  std::vector<float> dequantized_values(size);
+
+  const int len = 4;
+  static_assert(len == 4,"all weights should be set to the same rank as the convolution weights");
+  int len_per_dim[len];
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:228]: How about adding initialization code like this and for `indices` too?
```cxx
int len_per_dim[len] = { 0, }
```

Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:228]: And for above other codes too :)

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 11]
```diff
@@ -201,29 +389,58 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
...
+          std::vector<float> nudged_min(min.size());
+          std::vector<float> nudged_max(min.size());
+          std::vector<float> scaling_factor(min.size());
+          std::vector<int64_t> zp(min.size());
+
+          wquant_per_channel(circle_const, min, max, scaling_factor, zp, nudged_min, nudged_max);
+          wdequant_per_channel(circle_const, scaling_factor, nudged_min);
+
+          auto quantparam = std::make_unique<CircleQuantParam>();
+          for (std::vector<float>::size_type i = 0; i < min.size(); ++i)
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:409]: (optional) we can use `size_t` for `std::vector<float>::size_type`

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 12]
```diff
@@ -70,6 +70,198 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+}
+
+void wquant_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max,
+                        std::vector<float> &scaling_factor, std::vector<int64_t> &zp,
+                        std::vector<float> &nudged_min, std::vector<float> &nudged_max)
+{
+  const int32_t kMinScale = 0;
+  const int32_t kMaxScale = 255;
+
+  uint32_t size = node->size<loco::DataType::FLOAT32>();
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:180]: I think there is a precondition that `assert(node->dtype() ==loco::DataType::FLOAT32);`
can you please add this `assert` at the top of this function?

I think `wdequant_per_channel` will need this too.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 13]
```diff
@@ -201,29 +406,58 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
...
+          wdequant_per_channel(circle_const, scaling_factor, nudged_min);
+
+          auto quantparam = std::make_unique<CircleQuantParam>();
+          for (size_t i = 0; i < min.size(); ++i)
+          {
+            quantparam->min.push_back(nudged_min[i]);
+            quantparam->max.push_back(nudged_max[i]);
+            quantparam->scale.push_back(scaling_factor[i]);
+            quantparam->zerop.push_back(zp[i]);
+          }
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:432]: How about this change?
```suggestion
          quantparam->min = nudged_min;
          quantparam->max = nudged_max;
          quantparam->scale = scaling_factor;
          quantparam->zerop = zp;
```

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 14]
```diff
@@ -70,6 +70,210 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+int cal_offset(int *len_per_dim, int *indices)
+{
+  return indices[0] * len_per_dim[1] * len_per_dim[2] * len_per_dim[3] +
+         indices[1] * len_per_dim[2] * len_per_dim[3] + indices[2] * len_per_dim[3] + indices[3];
+}
+
+void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max)
+{
+  const int len = 4;
+  static_assert(len == 4, "all weights should be set to the same rank as the convolution weights");
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:133]: Could you tell me about the purpose of this assert? This assert will always pass I think.

Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:133]: I think it's a safe guard not to change `const int len = 4`

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 15]
```diff
@@ -70,6 +70,210 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+  int len_per_dim[len] = {
+      0,
+  };
+  int indices[len] = {
+      0,
+  };
+  int channel_dim_index;
+  int size;
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:141]: This suggestion is for static analysis warning issue
```suggestion
  int channel_dim_index{0};
  int size{0};
```

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 16]
```diff
@@ -70,6 +70,210 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
...
+  for (auto out : succs)
+  {
+    auto conv = dynamic_cast<CircleConv2D *>(out);
+    auto dw_conv = dynamic_cast<CircleDepthwiseConv2D *>(out);
+    auto tw_conv = dynamic_cast<CircleTransposeConv *>(out);
+    auto fc = dynamic_cast<CircleFullyConnected *>(out);
+
+    if (conv != nullptr && conv->filter() == node)
+    {
+      len_per_dim[0] = node->dim(0).value();
```
Comment (seanshpark) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:88]: What we discussed off-line about checking `node->rank() == 4`,
there is `bool is_weights(CircleNode *node)` method that does conditional check.
I think it would be better to check `node->rank() == 4` for `CircleConv2D ` in `is_weights()`...
@jinevening , what do you think?

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp [hunk 17]
```diff
@@ -201,29 +406,58 @@ struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<b
...
+          for (uint32_t i = 0; i < circle_const->size<loco::DataType::FLOAT32>(); i++)
+          {
+            auto data = circle_const->at<loco::DataType::FLOAT32>(i);
+            min = data < min ? data : min;
+            max = data > max ? data : max;
+          }
+          float scaling_factor;
+          int64_t zp;
+          float nudged_min;
+          float nudged_max;
```
Comment (mhs4670go) [inline compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:449]: ditto

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp (hunk 16 above)
Comment (jinevening) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:88]: > @jinevening , what do you think?

I didn't hear the off-line discussion.. I think the rank of CircleConv**2D** must be 4 by its definition. It should be checked when the node is created (and the rank should not be changed). AFAIK, there is no such guarantee, so it would be ok to add assertion to check the rank here.

Comment (jinevening) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:88]: > Adding assert(node->rank() == 4) is needed for safety.

I agree that it will make the code safer than before. Can you specify the line number where the assertion is inserted or make suggestions? It would be very helpful for @meejeong .

Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:88]: I think this can make clear?
1) add `if (node->rank() != 4) return false` in `is_weights()` method
2) add `assert(node->rank() == 4)` here, line 88.

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp (hunk 12 above)
Comment (meejeong) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:189]: I applied it line number ```184```

File: compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp (hunk 7 above)
Comment (seanshpark) [reply compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp:116]: Maybe you forgot to remove this change?

**Comment by seanshpark** [comment]
Looks OK with one comment :)
- https://github.com/Samsung/ONE/pull/2092#discussion_r438625638
